src/KeyManager.sol

```diff
@@ -33,62 +38,103 @@ abstract contract KeyManagerBase {
     // master private key
     function xPriv() internal virtual returns (bytes32);
 
-    // Key derivation for encryption
+    function _derivedPriv(address a, uint32 index) internal virtual returns (bytes32);
 
-    // Any contract in confidential mode can request a
-    // hardened derived key
-    function _derivedPriv(address a) private returns (bytes32) {
-        return keccak256(abi.encodePacked(a, xPriv()));
-    }
-
-    function derivedPriv() public returns (bytes32) {
-        return _derivedPriv(msg.sender);
-    }
+    // Key derivation for encryption
 
     // Because we are using hardened derivation, for each
     // contract we will need someone to sign it off chain
-    mapping(address => bytes) public derivedPub;
+    mapping(bytes32 => bytes) internal _derivedPub;
 
     function onchain_DeriveKey(
         address a,
         bytes memory dPub,
         // Signature from a valid kettle
-        bytes memory sig
+        bytes memory sig,
+        uint32 index
```

Comment:
index should go before the signature (the attestation), probably after the address - this is to indicate that the attestation is over the whole thing. As is it looks as if index was not attested to